client/verta/verta/pipeline/_pipeline_graph.py

```diff
@@ -0,0 +1,143 @@
+    """
+
+    def __init__(
+        self, steps: Union[List[PipelineStep], Set[PipelineStep], Tuple[PipelineStep]]
+    ):
+        self._steps = self._validate_steps(steps)
+
+    def __repr__(self) -> str:
+        return f"PipelineGraph steps:\n{self._format_steps()}"
+
+    def _format_steps(self) -> str:
+        """Format steps for improved readability in __repr__() function."""
+        return "\n".join([repr(s) for s in self._steps])
+
+    @property
+    def steps(self) -> Set[PipelineStep]:
+        return self._steps
+
+    @steps.setter
+    def steps(self, value):
+        raise AttributeError("can't set attribute 'steps'; please use set_steps()")
+
+    def set_steps(
+        self, steps: Union[List[PipelineStep], Set[PipelineStep], Tuple[PipelineStep]]
+    ) -> Set[PipelineStep]:
+        """Update the set of steps for this PipelineGraph to the provided value.
+
+        Parameters
+        ----------
+        steps : list, set, tuple of :class:`~verta.deployment.PipelineStep`
+            List, set, or tuple of all possible steps of the pipeline graph.
+            All options are converted to a set, so order is irrelevant and
+            duplicates are removed.
+
+        Returns
+        -------
+        set of :class:`~verta.deployment.PipelineStep`
+            The steps now set for this graph, if validation is successful.
+
+        Raises
+        ------
+        TypeError
+            If ``steps`` is not a set of PipelineStep objects.
+        """
+        self._steps = set(self._validate_steps(steps))
+        return self.steps
+
+    def _validate_steps(
+        self, steps: Union[List[PipelineStep], Set[PipelineStep], Tuple[PipelineStep]]
+    ) -> Set[PipelineStep]:
+        """Validate that the provided steps are a set of  PipelineStep objects.
+
+        Parameters
+        ----------
+        steps : list, set, or tuple of :class:`~verta.deployment.PipelineStep`
+            List, set, or tuple of steps provided by a user.
+
+        Returns
+        -------
+        set of :class:`~verta.deployment.PipelineStep`
+            The same set of steps if validation is successful.
+
+        Raises
+        ------
+        TypeError
+            If steps is not a set of PipelineStep objects.
+        """
+        if not isinstance(steps, (list, set, tuple)):
+            raise TypeError(
+                f"steps must be type list, set, or tuple, not {type(steps)}"
+            )
+        for step in steps:
+            if not isinstance(step, PipelineStep):
+                raise TypeError(
+                    f"individual steps of a PipelineGraph must be type"
+                    f" PipelineStep, not {type(step)}."
+                )
+            # throw an exception if any step's predecessors attr has been inappropriately mutated.
+            step._validate_predecessors(step.predecessors)
+        if len([s.name for s in steps]) is not len(set([s.name for s in steps])):
```

Comment:
`is` is **not** safe for comparing integers in Python! It happens to work (and only for small values) in CPython as an implementation detail, but it is never something we should rely on. `!=` is much safer.

```suggestion
        if len([s.name for s in steps]) != len(set([s.name for s in steps])):
```